### ISSUE TITLE
puppet-mgmtgraph should respect classes

### ISSUE THREAD

**Issue by purpleidea** [ffrank/puppet-mgmtgraph#1]
When we see classes in puppet like:

```
Package["foo"] -> Class["bar"]
```

and

```
class bar {
    file { 'a':
        ensure => present,
        content => "a",
    }
    file { 'b':
        ensure => present,
        content => "b",
    }
    file { 'c':
        ensure => present,
        content => "c",
    }
}
```

Is the equivalent raw graph something like:

```
Package["foo"] -> File["a"]
Package["foo"] -> File["b"]
Package["foo"] -> File["c"]
```

? I think this is the logical thing to do, but knowing what puppet *actually* does would be helpful. Maybe @tuxmea can chime in.

If so, we can replace the Class object with a "Noop" resource in mgmt, or just directly connect the edges. My uneducated preference would be to use the Noop resources so we could at least "see" the Class structure a bit.

**Comment by tuxmea** [comment]
Puppet does
```
Package['foo'] -> File['a'] -> File['b'] -> File['c']
```
I would prefer to have parallel execution.

**Comment by purpleidea** [comment]
@tuxmea Ohhh... Good point, that makes sense since puppet is not parallel. Would you agree with my suggested behaviour? And for this scenario:

```
Package["foo"] -> Class["bar"] -> Service["baz"]
```

should it produce:

```
Package["foo"] -> File["a"]
Package["foo"] -> File["b"]
Package["foo"] -> File["c"]
File["a"] -> Service["baz"]
File["b"] -> Service["baz"]
File["c"] -> Service["baz"]
```
(a diamond graph...)

in mgmt? I'm guessing puppet would do something like:

```
Package['foo'] -> File['a'] -> File['b'] -> File['c'] -> Service['baz']
```

Cheers

**Comment by tuxmea** [comment]
I agree with behavior from first post.
For the second example:
puppet would do as you described. But: Puppet 3 has an issue in case that class "bar" includes another class. As this nested class is not put into order. This is the containment problem I would try to prevent right from scratch.

Code example:
```
class 'bar' {
  file { 'a': }
  file { 'b': }
  include foobar
}
class foobar {
  file { 'c': }
}
package {'foo': ] -> class {'bar‘: } -> Service{'baz': }
```

Ordering must be the following (without foobar class):
```
Package['foo'] -> File['a]
Package['foo'] -> File['b']
File['a'] -> Service['baz']
File['b'] -> Service['baz']
```
Now we need to insert class foobar:
```
File['a'] -> File['c']
File['b'] -> File['c']
File['c'] -> Service['baz']
```
I wonder whether a strict 1-1 relation will prevent containment problems or whether we should see classes as some kind of sub ordering.

**Comment by purpleidea** [comment]
@tuxmea Your example looks good and the correct behaviour I would probably want to implement in mgmt (thanks to Felix for letting us have this discussion here...), although I didn't understand this comment:

> I wonder whether a strict 1-1 relation will prevent containment problems or whether we should see classes as some kind of sub ordering.

**Comment by tuxmea** [comment]
"strict 1-1 relation" is like mentioned in my last post.
"sub ordering" could be something like:

```
Package['foo'] -> SubEdge['bar']
SubEdge['bar:start'] -> File['a'] -> SubEdge['bar:end']
SubEdge['bar:start'] -> File[b'] -> SubEdge['bar:end']
SubEdge['bar:start'] -> SubEdge['foobar] -> SubEdge['bar:end']
SubEdge['foobar:start'] -> File['c'] -> SubEdge['foobar:end']
...
```
Hm. While typing this I already see that this seems to complex...

**Comment by ffrank** [comment]
Thanks to you both, but I think you're overthinking this.

First off, we cannot spawn relationships that Puppet won't generate. Please keep in mind that we're working off of the compiled catalog, not the manifest.

Second, the catalog *must* hold all needed edges, or Puppet itself would mess up dependencies. It's more a matter of understanding the actual graph structure under Puppet's hood. Then we can devise a way to transform it.

Also to clarify: Yes, Puppet will bring

```
Package['foo'] -> File['a'] -> File['b'] -> File['c']
```

into a deterministic *order*, but there are no dependency relationships between those file resources. So James' take of

```
Package["foo"] -> File["a"]
Package["foo"] -> File["b"]
Package["foo"] -> File["c"]
```

is closer to the mark, actually, since we only care for dependencies.

**Comment by ffrank** [comment]
Nailed it I think. One night of hacking precipitates yet another two weeks of blogging I guess :dizzy_face:

**Comment by purpleidea** [comment]
Amazingly short amount of code, but I'd believe it! What's a "whit"?

I really like this comment:

`* incidentally also adds support for stages (hah!)`

Great working, looking forward to reading about it and trying it out, and of course *ahem* a presentation at cfgmgmtcamp!

**Comment by ffrank** [comment]
The whit is a slightly confusing piece of Puppet internals with [hilarious commentary](https://github.com/puppetlabs/puppet/blob/f04567e4360a648bb54d878c998f260b31c542c9/lib/puppet/type/whit.rb). Apparently, these resources are generated internally to represent the "entry" and "exit" points of classes, defines, stages etc.

Probably using yet another private API now, but it's not a real problem. If push were to come to shove, people could always just freeze their translator Puppet gem at `4.5` or whatever.

**Comment by ffrank** [comment]
Should be resolved in `0.2.3`.
